utils/markdown.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package utils
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/yuin/goldmark"
+	"github.com/yuin/goldmark/ast"
+	"github.com/yuin/goldmark/extension"
+	astExt "github.com/yuin/goldmark/extension/ast"
+	"github.com/yuin/goldmark/renderer"
+	"github.com/yuin/goldmark/util"
+)
+
+func StripMarkdown(markdown string) string {
+	md := goldmark.New(
+		goldmark.WithExtensions(extension.GFM),
+		goldmark.WithRenderer(
+			renderer.NewRenderer(renderer.WithNodeRenderers(
+				util.Prioritized(newNotificationRenderer(), 500),
+			)),
+		),
+	)
+
+	var buf bytes.Buffer
+	if err := md.Convert([]byte(markdown), &buf); err != nil {
+		return ""
+	}
+
+	return strings.TrimSpace(buf.String())
+}
+
+type notificationRenderer struct {
+}
+
+func newNotificationRenderer() *notificationRenderer {
+	return &notificationRenderer{}
+}
+
+func (r *notificationRenderer) RegisterFuncs(reg renderer.NodeRendererFuncRegisterer) {
+	// block
+	reg.Register(ast.KindDocument, r.renderDefault)
+	reg.Register(ast.KindHeading, r.renderHeading)
+	reg.Register(ast.KindBlockquote, r.renderDefault)
+	reg.Register(ast.KindCodeBlock, r.renderCodeBlock)
+	reg.Register(ast.KindFencedCodeBlock, r.renderFencedCodeBlock)
+	reg.Register(ast.KindHTMLBlock, r.renderDefault)
+	reg.Register(ast.KindList, r.renderDefault)
+	reg.Register(ast.KindListItem, r.renderListItem)
+	reg.Register(ast.KindParagraph, r.renderParagraph)
+	reg.Register(ast.KindTextBlock, r.renderTextBlock)
+	reg.Register(ast.KindThematicBreak, r.renderDefault)
+
+	// inlines
+	reg.Register(ast.KindAutoLink, r.renderDefault)
+	reg.Register(ast.KindCodeSpan, r.renderDefault)
+	reg.Register(ast.KindEmphasis, r.renderDefault)
+	reg.Register(ast.KindImage, r.renderDefault)
+	reg.Register(ast.KindLink, r.renderDefault)
+	reg.Register(ast.KindRawHTML, r.renderDefault)
+	reg.Register(ast.KindText, r.renderText)
+	reg.Register(ast.KindString, r.renderString)
+
+	// table
+	reg.Register(astExt.KindTable, r.renderTable)
+
+	// strikethrough
+	reg.Register(astExt.KindStrikethrough, r.renderDefault)
+}
+
+// renderDefault renderer function to renderDefault without changes
+func (r *notificationRenderer) renderDefault(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	return ast.WalkContinue, nil
+}
+
+func (r *notificationRenderer) renderHeading(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	if !entering {
+		if node.NextSibling() != nil {
+			_, _ = w.WriteString(" ")
+		}
+	}
+	return ast.WalkContinue, nil
+}
+
+func (r *notificationRenderer) renderListItem(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	if entering {
+		fc := node.FirstChild()
+		if fc != nil {
+			if _, ok := fc.(*ast.TextBlock); !ok {
+				_ = w.WriteByte(' ')
+			}
+		}
```

Comment:
Can you explain why this is needed? Removing this does not fail any test case. In that case, `renderHeading`, `renderListItem` and `renderParagraph` can all become the same method.

Reply:
Actually I just copy-paste from html renderer, yeah those three function can be wrap into only one function, thanks.